### PR TITLE
Update IndexingGenerator.next() function for removal of ++ operator

### PR DESCRIPTION
**Pull request by nielsandriesse**
Related to: https://github.com/apple/swift-evolution/blob/master/proposals/0004-remove-pre-post-inc-decrement.md

### CODE REVIEW

**Comment by nielsandriesse** [comment]
Comparison of SIL: https://gist.github.com/nielsandriesse/556b236ed7aed8763f33

File: stdlib/public/core/Collection.swift [hunk 1]
```diff
@@ -101,8 +101,10 @@ public struct IndexingGenerator<Elements : Indexable>
   ///
   /// - Requires: No preceding call to `self.next()` has returned `nil`.
   public mutating func next() -> Elements._Element? {
-    return _position == _elements.endIndex
-    ? .None : .Some(_elements[_position++])
+    guard _position != _elements.endIndex else { return .None }
```
Comment (TomasLinhart) [inline stdlib/public/core/Collection.swift]: Wouldn't be `if` better and reverse the condition? The guard seems to be good for unwrapping but just for checking booleans it looks weird.

File: stdlib/public/core/Collection.swift [hunk 2]
```diff
@@ -101,8 +101,10 @@ public struct IndexingGenerator<Elements : Indexable>
...
+    let element: Elements._Element? = .Some(_elements[_position])
+    _position = _position.successor()
```
Comment (gribozavr) [inline stdlib/public/core/Collection.swift]: Should be `_position._successorInPlace()`.

File: stdlib/public/core/Collection.swift (hunk 1 above)
Comment (gribozavr) [inline stdlib/public/core/Collection.swift]: `return nil`

File: stdlib/public/core/Collection.swift [hunk 3]
```diff
@@ -101,8 +101,10 @@ public struct IndexingGenerator<Elements : Indexable>
   ///
   /// - Requires: No preceding call to `self.next()` has returned `nil`.
   public mutating func next() -> Elements._Element? {
-    return _position == _elements.endIndex
-    ? .None : .Some(_elements[_position++])
+    guard _position != _elements.endIndex else { return .None }
+    let element: Elements._Element? = .Some(_elements[_position])
```
Comment (gribozavr) [inline stdlib/public/core/Collection.swift]: `let element = _elements[_position]` ?

**Comment by nielsandriesse** [comment]
Performance seems identical (optimized build): https://gist.github.com/nielsandriesse/3ff2a9eaca6ec4d386f6

**Comment by nielsandriesse** [comment]
Ran performance tests again after last commit. Results are still identical.

**Comment by gribozavr** [comment]
@nadavrot This PR needs attention from your team.

**Comment by nadavrot** [comment]
If this patch does not degrade the performance of next() then I have no objections. We'll monitor the performance of our tests post-commit. Thanks @gribozavr.

**Comment by gribozavr** [comment]
Thanks, @nadavrot!  Merging.

**Comment by nielsandriesse** [comment]
Considering the pending removal of these operators, I filed a bug for this optimization case: https://bugs.swift.org/browse/SR-213

**Pull request merged**